cmd/store.go

```diff
@@ -31,64 +35,133 @@ func (s *store) Get(IP string) (string, error) {
 	return "", fmt.Errorf("Unable to find role for IP %s", IP)
 }
 
-// OnAdd is called when a pod is added.
-func (s *store) OnAdd(obj interface{}) {
-	pod, ok := obj.(*api.Pod)
-	if !ok {
-		log.Errorf("Expected Pod but OnAdd handler received %+v", obj)
-		return
+func (s *store) AddRoleToIP(pod *api.Pod, role string) {
+	s.mutex.Lock()
+	s.rolesByIP[pod.Status.PodIP] = role
+	s.mutex.Unlock()
+}
+
+func (s *store) AddNamespaceToIP(pod *api.Pod) {
+	namespace := pod.GetNamespace()
+	s.mutex.Lock()
+	s.namespaceByIP[pod.Status.PodIP] = namespace
+	s.mutex.Unlock()
+}
+
+func (s *store) DeleteIP(ip string) {
+	s.mutex.Lock()
+	delete(s.rolesByIP, ip)
+	delete(s.namespaceByIP, ip)
+	s.mutex.Unlock()
+}
+
+// AddRoleToNamespace takes a role name and adds it to our internal state
+func (s *store) AddRoleToNamespace(namespace string, role string) {
+	// Make sure to add the full ARN of roles to ensure string matching works
+	roleARN := s.iam.roleARN(role)
+
+	ar := s.rolesByNamespace[namespace]
+	if ar == nil {
+		ar = []string{}
 	}
```

Comment:
This block is unnecessary - appending to an empty slice is ok - it will allocate memory/etc automatically.